src/integritysetup/integritysetup-generator.c

```diff
@@ -0,0 +1,224 @@
+                if (!out)
+                        return log_oom();
+
+                token = strtok_r(options, ",", &save_ptr);
+                while (token != NULL) {
+                        option(out, token);
+                        token = strtok_r(NULL, ",", &save_ptr);
+                }
+
+                r = fclose(out);
+                if (r != 0)
+                        return log_error_errno(r, "fclose failure on opened memstream: %m");
+
+                if (len > 0) {
+                        *options_formatted = buff;
+                }
+                return 0;
+        }
+        return -EINVAL;
+}
+
+static int create_disk(
+                const char *name,
+                const char *device,
+                char *options) {
+
+        _cleanup_free_ char *n = NULL, *dd = NULL, *hd = NULL, *e = NULL, *to = NULL, *dev_require = NULL,
+                            *d_escaped = NULL, *hu_escaped = NULL, *name_escaped = NULL, *options_convert = NULL;
+        _cleanup_fclose_ FILE *f = NULL;
+        int r;
+        char *dmname = NULL;
+
+        assert(name);
+        assert(device);
+
+        name_escaped = specifier_escape(name);
+        if (!name_escaped)
+                return log_oom();
+
+        e = unit_name_escape(name);
+        if (!e)
+                return log_oom();
+
+        r = unit_name_build("systemd-integritysetup", e, ".service", &n);
+        if (r < 0)
+                return log_error_errno(r, "Failed to generate unit name: %m");
+
+        r = unit_name_from_path(device, ".device", &dd);
+        if (r < 0)
+                return log_error_errno(r, "Failed to generate unit name: %m");
+
+        r = generator_open_unit_file(arg_dest, NULL, n, &f);
+        if (r < 0)
+                return r;
+
+        if (options) {
+                r = options_string(options, &options_convert);
+                if (r < 0)
+                        return r;
+        }
+
+        fprintf(f,
+                "[Unit]\n"
+                "Description=Integrity Setup for %%I\n"
+                "Documentation=man:integritytab(5) man:systemd-integritysetup-generator(8) man:systemd-integritysetup@.service(8)\n"
+                "SourcePath=%s\n"
+                "DefaultDependencies=no\n"
+                "IgnoreOnIsolate=true\n"
+                "After=systemd-udevd-kernel.socket\n"
+                "Before=blockdev@dev-mapper-%%i.target\n"
+                "Wants=blockdev@dev-mapper-%%i.target\n"
+                "Conflicts=umount.target\n"
+                "Before=integritysetup.target\n"
+                "BindsTo=%s\n"
+                "After=%s\n"
+                "Before=umount.target\n",
+                arg_integritytab,
+                dd, dd);
+
+        fprintf(f,
+                "\n"
+                "[Service]\n"
+                "Type=oneshot\n"
+                "RemainAfterExit=yes\n"
+                "TimeoutSec=0\n"
+                "ExecStart=%s open '%s' '%s' %s\n"
```

Comment:
maybe add `ExecStart=%s close --deferred ...` so that there's no need for the explicit stop?

Reply:
The `integritysetup` tool does not support the `--deferred` option, or perhaps I'm missing some systemd services feature?

Reply:
it does support it - I've added it... https://man7.org/linux/man-pages/man8/integritysetup.8.html

Reply:
> it does support it - I've added it... https://man7.org/linux/man-pages/man8/integritysetup.8.html

Well upstream repo. does show your commit for Aug 2020, but the first release with it looks like is was included is 2.4.0 which was done Aug, 2021. It's not in Fedora 34 which I'm using for development, thus the reason for my confusion.

From looking at the updated man page you reference, this defers the close to the last open.  I'm assuming this has to be run after some process has opened the device mapper target?

If that is true then I'm not sure how this would work as this service unit gets run when the block device shows up which is before the file systems get mounted.

Reply:
yes indeed, it can't be used here, never mind